### ISSUE TITLE
Devise a way to test remote APIs

### ISSUE THREAD

**Issue by Minoru** [newsboat/newsboat#190]
This is a continuation of a discussion started in #187.

The problem: if Newsboat starts to behave wrong with some remote API, we can't easily figure out which end—Newsboat or the API—is at fault, and what exactly changed.

I wouldn't retell the proposed solutions here; go read the aforementioned issue, it's just 2 comments.

**Comment by Minoru** [comment]
Responding to [this comment](https://github.com/newsboat/newsboat/pull/187#issuecomment-388553412) by @noctux.

> Most APIs use utils::retrieve_url. We could use objcopy to weaken that symbol in the .o file when linking to the tests […]

That's a viable way to get existing code under test, but it's hardly maintainable. I didn't yet look at the code to figure out how feasible it is, but here's an idea (inspired by Haskell, I guess): split logic from I/O, and write unit tests just for logic. We'll get something like 90% coverage without any linker-level hackery. The downside to that approach is "code fragmentation": instead of long functions that intersperse logic with I/O, we'll have multiple methods that implement chunks of logic, and a method that ties them together. So maybe that wouldn't work that well for APIs that require multiple different calls.

**Comment by noctux** [comment]
Well, Haskell is lazy with class, but without classes (ok, I'm sorry for that bad pun. But only a tiny bit :p) and with too much style and syntactic sugar.
So I would not fancy implementing a monadic I/O interface in C++, which sounds like a lot of pain.

Furthermore, I think we could do better than the proposal of having an "I/O-centic" toplevel method that calls several pure ones.
How about introducing an "I/O-Manager" class/singleton/*, which offers the actual I/O-interfaces.
That way, we could trivially subclass it in our tests and thereby stub the network-related functionality, and test against known api-responses, which admittedly adds a bit of overhead but looks like an ideomatic C++ solution to me.

**Comment by Minoru** [comment]
> So I would not fancy implementing a monadic I/O interface in C++

Oh, I wouldn't either! But luckily I wasn't suggesting that :) All I'm saying is "we should put as much stuff as possible into pure functions, and write unit tests for those".

> How about introducing an "I/O-Manager" class/singleton/*, which offers the actual I/O-interfaces.

Weren't you pointing the downsides of this approach in your earlier comment in #187? I mean this part:

> But spawning the server, asserting that it's there, communicating which trace is relevant, etc. sounds like a major hassle.

"I/O-Manager" as a class would remove the first two downsides, but third is still an issue. Is it time to look into mocking or faking frameworks? Writing stubs is their job, as far as I understand.

Stepping away a bit, I want to state the following explicitly: if you have some approach you *want* to implement, just go for it. You'll have fun, and we'll have some evidence for and against the approach you choose. Design is important, but there's only so much we can achieve by just talking :)

**Comment by noctux** [comment]
> Weren't you pointing the downsides of this approach in your earlier comment in #187? I mean this part [...]

No, back then, I was thinking about actually spawning a server(-mock) (e.g. some perl script) during the tests, which would be accessed using `retrieve_url/libcurl` via an actual TCP-socket.

The I/O-manager I'm talking about above instead would represent an abstract interface to network-io containing functions such as `retrieve_url`.
Using subclasses and overriding, we could then stub these functions with implementations that check certain request properties and hand out predefined results (e.g., from traces against the original implementation), keeping both the code as well as the mock within the testcase itself, solving all three problems. The downside is that this solution would entail a bit of refactoring and a minor performance loss.

> if you have some approach you want to implement, just go for it

Good to know, thanks. However, when doing so, I would like a durable solution :) We'll see

**Comment by Minoru** [comment]
After a bit of thinking, I've become less enamored by "pure functions" approach I described. It will lead to either:
1. many static functions (which is un-idiomatic), or
2. complicated tests which need to maneuver the object into appropriate state (which kind of defeats the purpose, and ideologically isn't much different from "I/O-Manager" approach).

Faking the server indeed seems the best solution.

Since there will be just one "I/O-manager" implementation in Newsboat (the rest will be in tests), compiler should be able to de-virtualize the calls, so we probably won't even lose any performance at all \o/
